pkg/node/node.go

```diff
@@ -101,105 +103,261 @@ func createAddressFilter(
 	}
 }
 
-// GenerateSignerForKeep generates a new threshold signer with ECDSA key pair. The
-// public key is a public key of the signing group. It publishes the public key
-// to the keep. It uses keep address as unique signing group identifier.
+// GenerateSignerForKeep generates a new threshold signer with ECDSA key pair
+// and submits the public key to the on-chain keep.
+//
+// The attempt for generating signer is retried on failure until the provided
+// context is done.
 func (n *Node) GenerateSignerForKeep(
 	ctx context.Context,
 	operatorPublicKey *operator.PublicKey,
 	keepAddress common.Address,
-	keepMembersIDs []tss.MemberID,
+	members []common.Address,
 ) (*tss.ThresholdSigner, error) {
 	memberID := tss.MemberIDFromPublicKey(operatorPublicKey)
 
-	signer, err := tss.GenerateThresholdSigner(
-		ctx,
-		keepAddress.Hex(),
-		memberID,
-		keepMembersIDs,
-		uint(len(keepMembersIDs)-1),
-		n.networkProvider,
-		n.tssParamsPool.get(),
-	)
-	if err != nil {
-		return nil, fmt.Errorf("failed to generate threshold signer: [%v]", err)
+	logger.Infof("Getting new key-generation pre-parameters from the pool")
+	preParamsBox := params.NewBox(n.tssParamsPool.get())
+	logger.Infof("Received new key-generation parameters from the pool")
+
+	attemptCounter := 0
+	for {
+		attemptCounter++
+
+		logger.Infof(
+			"signer generation for keep [%s]; attempt [%v]",
+			keepAddress.String(),
+			attemptCounter,
+		)
+
+		// If we are re-attempting the key generation, pre-parameters in the box
+		// could be destroyed because they were shared with other members.
+		// In this case, we need to re-generate them.
+		if preParamsBox.IsEmpty() {
+			logger.Infof("Getting new key-generation pre-parameters from the pool")
+			preParamsBox = params.NewBox(n.tssParamsPool.get())
+			logger.Infof("Received new key-generation parameters from the pool")
```

Comment:
Same here as in the comment above.

Reply:
Done in https://github.com/keep-network/keep-ecdsa/pull/357/commits/e8b95aea03fd3e029a5ee5120110f90f9dfbdd5f